### DISCUSSION TITLE
Setting x:DataType on Selector when using CompileBindings problem

### DISCUSSION THREAD

**Discussion by MihaMarkic** [AvaloniaUI/Avalonia#8688] [General]
I have an MVVM backed UserControl like one below. It works fine when using reflection binding, however it would throw a runtime exception "Can't cast MasterViewModel to ChildViewModel" at runtime when using CompileBindings.
Seems like CompileBinding is using wrong DataContext (one from UserControl instead one from Item in ItemsControl).
Is this a known bug with CompileBindings? Worth investigating?

```xaml
<UserControl x:DataType="vm:MasterViewModel">
  <UserContro.Styles>
     <Style Selector="ItemsControl > ContentPresenter" x:DataType="vm:ChildViewModel">
       <Setter Property ... Value="{Binding ValueOnChild}" />
     </Style>
  </UserContro.Styles>
  <ItemsControl Items="{Binding Children}" />
</UserControl>
```

**Comment by maxkatz6** [comment]
> Seems like CompileBinding is using wrong DataContext

No, it is expecting ChildViewModel just like you asked it to do.
But actual data context was wrong - MasterViewModel.

Can it be the case, that content presenter was added to the visual tree without item's data context?

**Reply by MihaMarkic** [reply]
Yes, probably I've made a mistake somewhere as I tried to reproduce it in a simple sample and it seems working as expected.
